src/compiler/utilities.ts

```diff
@@ -4033,7 +4034,15 @@ namespace ts {
         const filePath = getNormalizedAbsolutePath(fileName, host.getCurrentDirectory());
         const relativePath = getRelativePathToDirectoryOrUrl(dir, filePath, dir, getCanonicalFileName, /*isAbsolutePathAnUrl*/ false);
         const extensionless = removeFileExtension(relativePath);
-        return referencePath ? ensurePathIsNonModuleName(extensionless) : extensionless;
+        if (referencePath) {
+            return ensurePathIsNonModuleName(extensionless);
+        }
+        const rootPkgName = host.getCompilerOptions().bundledPackageName || "";
+        const newPath = combinePaths(rootPkgName, extensionless);
+        if (endsWith(newPath, "/index")) {

```

Comment:
We should probably only do this if `host.getCompilerOptions().bundledPackageName || "";
        const newPath = combinePaths(rootPkgName, extensionless);` exists and `getEmitModuleResolutionKind` is `NodeJs`, so as to not unintentionally change any other behaviors.

Reply:
Naive question: how do we tell if `host.getCompilerOptions().bundledPackageName || ""` exists?

Reply:
I mean, we don't _have_ to check for the `""` and could just check the compiler option - that `""` is a noop for `combinePaths` since it doesn't take `undefined`. (It's still falsy tho, so you can just check falsyness)

Reply:
Added, thanks!